numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +154,188 @@ def potential_fn(z_gibbs, z_hmc):
         z = {**z_gibbs, **hmc_state.z}
 
         return HMCGibbsState(z, hmc_state, rng_key)
+
+
+def _discrete_step(rng_key, z_discrete, pe, potential_fn, idx, support_size):
+    # idx: current index of `z_discrete_flat` to update
+    # support_size: support size of z_discrete at the index idx
+
+    # get z proposal
+    z_discrete_flat, unravel_fn = ravel_pytree(z_discrete)
+
+    # XXX: we can't vmap potential_fn over all proposals and sample from the conditional
+    # categorical distribution because support_size is a traced value, i.e. its value
+    # might change across different discrete variables;
+    # so here we will loop over all proposals and use an online scheme to sample from
+    # the conditional categorical distribution
+    def body_fn(i, val):
+        rng_key, z, pe, weight_logsumexp = val
+        rng_key, rng_accept = random.split(rng_key)
+        proposal = jnp.where(i >= z_discrete_flat[idx], i + 1, i)
```

Comment:
i don't understand this where is the randomness coming from?

Reply:
Based on the weights, we will decide if we will accept the proposal or not (`random.bernoulli(rng_accept, transition_prob)` below). For example, assume we only have 1 variable, support_size=4 and the probs are, 0.3, 0.1, 0.2, 0.4; idx=0: Then starting at 0, we will decide if we move to 1 based on the transition_prob 0.1/(0.3 + 0.1) = 0.25. Then we will decide if we move to 2 based on the transition prob `0.2 / (0.2 + 0.25) = 4/9`. Finally, we will decide if we move to 3 based on the transition prob `0.4 / (0.4 + 4.9)`. Is it reasonable?

Reply:
doesn't this mean you have a 75% chance of staying at 0?

Reply:
Yes, that is for first step. If it stays at 0, then for the second step, you will have 5/9 chance to stay at 0.

Reply:
Opps, chance to stay at 0 for the second step is 2/3. I just edited the numbers.

Reply:
So chance to stay at 0 until the end is: 3/4 * 2/3 * 3/5 = 0.3

Reply:
👍